compression_verifier.go

```diff
@@ -0,0 +1,238 @@
+		"table": table,
+	}).Info("decompressing table data before verification")
+
+	tableCompression := c.tableColumnCompressions[table]
+
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := getRows(db, schema, table, pkColumn, columns, pks)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	// Decompress applicable columns and hash the resulting column values for comparison
+	resultSet := make(map[uint64][]byte)
+	for rows.Next() {
+		rowData, err := ScanByteRow(rows, len(columns)+1)
+		if err != nil {
+			return nil, err
+		}
+
+		pk, err := strconv.ParseUint(string(rowData[0]), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decompress the applicable columns and then hash them together
+		// to create a fingerprint. decompressedRowData contains a map of all
+		// the non-compressed columns and associated decompressed values by the
+		// index of the column
+		decompressedRowData := [][]byte{}
+		for idx, column := range columns {
+			if algorithm, ok := tableCompression[column.Name]; ok {
+				// rowData contains the result of "SELECT pkColumn, * FROM ...", so idx+1 to get each column
+				decompressedColData, err := c.Decompress(table, column.Name, algorithm, rowData[idx+1])
+				if err != nil {
+					return nil, err
+				}
+				decompressedRowData = append(decompressedRowData, decompressedColData)
+			} else {
+				decompressedRowData = append(decompressedRowData, rowData[idx+1])
+			}
+		}
+
+		// Hash the data of the row to be added to the result set
+		decompressedRowHash, err := c.HashRow(decompressedRowData)
+		if err != nil {
+			return nil, err
+		}
+
+		resultSet[pk] = decompressedRowHash
+	}
+
+	metrics.Gauge("compression_verifier_decompress_rows", float64(len(resultSet)), []MetricTag{}, 1.0)
+	logrus.WithFields(logrus.Fields{
+		"tag":   "compression_verifier",
+		"rows":  len(resultSet),
+		"table": table,
+	}).Debug("decompressed rows will be compared")
+
+	return resultSet, nil
+}
+
+// Decompress will apply the configured decompression algorithm to the configured columns data
+func (c *CompressionVerifier) Decompress(table, column, algorithm string, compressed []byte) ([]byte, error) {
+	var decompressed []byte
+	switch strings.ToUpper(algorithm) {
+	case CompressionSnappy:
+		return snappy.Decode(decompressed, compressed)
+	default:
+		return nil, UnsupportedCompressionError{
+			table:     table,
+			column:    column,
+			algorithm: algorithm,
+		}
+	}
+
+}
+
+// HashRow will fingerprint the non-primary columns of the row to verify data equality
+func (c *CompressionVerifier) HashRow(decompressedRowData [][]byte) ([]byte, error) {
+	if len(decompressedRowData) == 0 {
+		return nil, errors.New("Row data to fingerprint must not be empty")
+	}
+
+	hash := md5.New()
+	var rowFingerprint []byte
+	for _, colData := range decompressedRowData {
+		rowFingerprint = append(rowFingerprint, colData...)
+	}
+
+	hash.Write(rowFingerprint)
+	return []byte(hex.EncodeToString(hash.Sum(nil))), nil
```

Comment:
Note to other reviewers: this `Sum()` method is not the same as `md5.Sum()`

We're actually working with a `hash.Hash` implementation. See the example: https://golang.org/pkg/crypto/md5/#example_New